src/rust/engine/nailgun/src/client.rs

```diff
@@ -0,0 +1,152 @@
+#![deny(
+  clippy::all,
+  clippy::default_trait_access,
+  clippy::expl_impl_clone_on_copy,
+  clippy::if_not_else,
+  clippy::needless_continue,
+  clippy::unseparated_literal_suffix,
+// TODO: Falsely triggers for async/await:
+//   see https://github.com/rust-lang/rust-clippy/issues/5360
+// clippy::used_underscore_binding
+)]
+// It is often more clear to show that nothing is being moved.
+#![allow(clippy::match_ref_pats)]
+// Subjective style.
+#![allow(
+  clippy::len_without_is_empty,
+  clippy::redundant_field_names,
+  clippy::too_many_arguments
+)]
+// Default isn't as big a deal as people seem to think it is.
+#![allow(clippy::new_without_default, clippy::new_ret_no_self)]
+// Arc<Mutex> can be more clear than needing to grok Orderings:
+#![allow(clippy::mutex_atomic)]
+
+use nails::execution::{stream_for, ChildInput, ChildOutput, ExitCode};
+use nails::Config;
+use tokio::net::TcpStream;
+
+use std::io;
+use std::net::Ipv4Addr;
+use tokio::io::AsyncWriteExt;
+
+use futures::channel::{mpsc, oneshot};
+use futures::{future, FutureExt, SinkExt, Stream, StreamExt};
+use log::debug;
+
+pub enum NailgunClientError {
+  PreConnect(String),
+  PostConnect(String),
+  ExplicitQuit,
+}
+
+async fn handle_client_output(
+  mut stdio_read: impl Stream<Item = ChildOutput> + Unpin,
+) -> Result<(), io::Error> {
+  let mut stdout = tokio::io::stdout();
+  let mut stderr = tokio::io::stderr();
+  while let Some(output) = stdio_read.next().await {
+    match output {
+      ChildOutput::Stdout(bytes) => stdout.write_all(&bytes).await?,
+      ChildOutput::Stderr(bytes) => stderr.write_all(&bytes).await?,
+      ChildOutput::Exit(_) => {
+        // NB: We ignore exit here and allow the main thread to handle exiting.
+        break;
+      }
+    }
+  }
+  Ok(())
+}
+
+async fn handle_client_input(mut stdin_write: mpsc::Sender<ChildInput>) -> Result<(), io::Error> {
+  use nails::execution::send_to_io;
+  let mut stdin = stream_for(tokio::io::stdin());
+  while let Some(input_bytes) = stdin.next().await {
+    stdin_write
+      .send(ChildInput::Stdin(input_bytes?))
+      .await
+      .map_err(send_to_io)?;
+  }
+  stdin_write
+    .send(ChildInput::StdinEOF)
+    .await
+    .map_err(send_to_io)?;
+  Ok(())
+}
+
+async fn client_execute_helper(
+  port: u16,
+  command: String,
+  args: Vec<String>,
+  env: Vec<(String, String)>,
+) -> Result<i32, NailgunClientError> {
+  use nails::execution::{child_channel, Command};
+
+  let working_dir =
+    std::env::current_dir().map_err(|e| NailgunClientError::PreConnect(e.to_string()))?;
+
+  let config = Config::default();
+  let command = Command {
+    command,
+    args,
+    env,
+    working_dir,
+  };
+
+  let (stdio_write, stdio_read) = child_channel::<ChildOutput>();
+  let (stdin_write, stdin_read) = child_channel::<ChildInput>();
+
+  let output_handler = tokio::spawn(handle_client_output(stdio_read));
+  let _input_handler = tokio::spawn(handle_client_input(stdin_write));
```

Comment:
When is this future going to execute? Will be immediately? Do we need to `.await?` it at all?

Reply:
This was copied from @stuhood 's example code in his `nails` crate, where the spawned handler was also left untouched. I believe it's fine if nothing explicitly waits on the input handler and it gets shut down when the executor shuts down when the client `pants` process exits.